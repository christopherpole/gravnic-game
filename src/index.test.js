import {
  MOVE_UP,
  MOVE_RIGHT,
  MOVE_DOWN,
  MOVE_LEFT,
  convertTilesToGameState,
  calulateNextGameState,
  changeGravityDirection,
} from './index';
import Entity from './entity';

const testTiles = [
  {
    position: 1,
    selectedTileId: Entity.NONE,
  },
  {
    position: 2,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 3,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 4,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 5,
    selectedTileId: Entity.NONE,
  },
  {
    position: 6,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 7,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 8,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 9,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 10,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 11,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 12,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 13,
    selectedTileId: Entity.BLOCK,
  },
  {
    position: 14,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 15,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 16,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 17,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 18,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 19,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 20,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 21,
    selectedTileId: Entity.NONE,
  },
  {
    position: 22,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 23,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 24,
    selectedTileId: Entity.FLOOR,
  },
  {
    position: 25,
    selectedTileId: Entity.NONE,
  },
];

describe('Gravnic', () => {
  describe('convertTilesToGameState()', () => {
    it('Converts the tiles into a starting game state', () => {
      const gameState = convertTilesToGameState(testTiles);

      expect(gameState).toMatchSnapshot();
    });
  });

  describe('calulateNextGameState()', () => {
    it('Should return the next game state when moving entities up', () => {
      const gameState = calulateNextGameState(convertTilesToGameState(testTiles), MOVE_UP);

      expect(gameState).toMatchSnapshot();
    });

    it('Should return the next game state when moving entities right', () => {
      const gameState = calulateNextGameState(convertTilesToGameState(testTiles), MOVE_RIGHT);

      expect(gameState).toMatchSnapshot();
    });

    it('Should return the next game state when moving entities down', () => {
      const gameState = calulateNextGameState(convertTilesToGameState(testTiles), MOVE_DOWN);

      expect(gameState).toMatchSnapshot();
    });

    it('Should return the next game state when moving entities left', () => {
      const gameState = calulateNextGameState(convertTilesToGameState(testTiles), MOVE_LEFT);

      expect(gameState).toMatchSnapshot();
    });

    it('Should return false if the entities cannot move any further', () => {
      let gameState = calulateNextGameState(convertTilesToGameState(testTiles), MOVE_LEFT);
      gameState = calulateNextGameState(gameState, MOVE_LEFT);
      gameState = calulateNextGameState(gameState, MOVE_LEFT);

      expect(gameState).toBe(false);
    });
  });

  describe('changeGravityDirection()', () => {
    it('Should return an array of all game states from the move', () => {
      let gameStates = changeGravityDirection(convertTilesToGameState(testTiles), MOVE_DOWN);
      expect(gameStates).toMatchSnapshot();

      gameStates = changeGravityDirection(gameStates[gameStates.length - 1], MOVE_UP);
      expect(gameStates).toMatchSnapshot();
    });
  });
});
