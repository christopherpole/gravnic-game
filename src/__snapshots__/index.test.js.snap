// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gravnic calulateNextGameState() Should return the next game state when moving entities down 1`] = `
Array [
  Array [
    Object {
      "entityId": 0,
      "id": 0,
    },
    Object {
      "entityId": 1,
      "id": 1,
    },
    Object {
      "entityId": 1,
      "id": 2,
    },
    Object {
      "entityId": 1,
      "id": 3,
    },
    Object {
      "entityId": 0,
      "id": 4,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 5,
    },
    Object {
      "entityId": 1,
      "id": 6,
    },
    Object {
      "entityId": 1,
      "id": 7,
    },
    Object {
      "entityId": 1,
      "id": 8,
    },
    Object {
      "entityId": 1,
      "id": 9,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 10,
    },
    Object {
      "entityId": 1,
      "id": 11,
    },
    Object {
      "entityId": 1,
      "id": 12,
    },
    Object {
      "entityId": 1,
      "id": 13,
    },
    Object {
      "entityId": 1,
      "id": 14,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 15,
    },
    Object {
      "entityId": 1,
      "id": 16,
    },
    Object {
      "entityId": 2,
      "id": 17,
    },
    Object {
      "entityId": 1,
      "id": 18,
    },
    Object {
      "entityId": 1,
      "id": 19,
    },
  ],
  Array [
    Object {
      "entityId": 0,
      "id": 20,
    },
    Object {
      "entityId": 1,
      "id": 21,
    },
    Object {
      "entityId": 1,
      "id": 22,
    },
    Object {
      "entityId": 1,
      "id": 23,
    },
    Object {
      "entityId": 0,
      "id": 24,
    },
  ],
]
`;

exports[`Gravnic calulateNextGameState() Should return the next game state when moving entities left 1`] = `
Array [
  Array [
    Object {
      "entityId": 0,
      "id": 0,
    },
    Object {
      "entityId": 1,
      "id": 1,
    },
    Object {
      "entityId": 1,
      "id": 2,
    },
    Object {
      "entityId": 1,
      "id": 3,
    },
    Object {
      "entityId": 0,
      "id": 4,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 5,
    },
    Object {
      "entityId": 1,
      "id": 6,
    },
    Object {
      "entityId": 1,
      "id": 7,
    },
    Object {
      "entityId": 1,
      "id": 8,
    },
    Object {
      "entityId": 1,
      "id": 9,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 10,
    },
    Object {
      "entityId": 2,
      "id": 11,
    },
    Object {
      "entityId": 1,
      "id": 12,
    },
    Object {
      "entityId": 1,
      "id": 13,
    },
    Object {
      "entityId": 1,
      "id": 14,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 15,
    },
    Object {
      "entityId": 1,
      "id": 16,
    },
    Object {
      "entityId": 1,
      "id": 17,
    },
    Object {
      "entityId": 1,
      "id": 18,
    },
    Object {
      "entityId": 1,
      "id": 19,
    },
  ],
  Array [
    Object {
      "entityId": 0,
      "id": 20,
    },
    Object {
      "entityId": 1,
      "id": 21,
    },
    Object {
      "entityId": 1,
      "id": 22,
    },
    Object {
      "entityId": 1,
      "id": 23,
    },
    Object {
      "entityId": 0,
      "id": 24,
    },
  ],
]
`;

exports[`Gravnic calulateNextGameState() Should return the next game state when moving entities right 1`] = `
Array [
  Array [
    Object {
      "entityId": 0,
      "id": 0,
    },
    Object {
      "entityId": 1,
      "id": 1,
    },
    Object {
      "entityId": 1,
      "id": 2,
    },
    Object {
      "entityId": 1,
      "id": 3,
    },
    Object {
      "entityId": 0,
      "id": 4,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 5,
    },
    Object {
      "entityId": 1,
      "id": 6,
    },
    Object {
      "entityId": 1,
      "id": 7,
    },
    Object {
      "entityId": 1,
      "id": 8,
    },
    Object {
      "entityId": 1,
      "id": 9,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 10,
    },
    Object {
      "entityId": 1,
      "id": 11,
    },
    Object {
      "entityId": 1,
      "id": 12,
    },
    Object {
      "entityId": 2,
      "id": 13,
    },
    Object {
      "entityId": 1,
      "id": 14,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 15,
    },
    Object {
      "entityId": 1,
      "id": 16,
    },
    Object {
      "entityId": 1,
      "id": 17,
    },
    Object {
      "entityId": 1,
      "id": 18,
    },
    Object {
      "entityId": 1,
      "id": 19,
    },
  ],
  Array [
    Object {
      "entityId": 0,
      "id": 20,
    },
    Object {
      "entityId": 1,
      "id": 21,
    },
    Object {
      "entityId": 1,
      "id": 22,
    },
    Object {
      "entityId": 1,
      "id": 23,
    },
    Object {
      "entityId": 0,
      "id": 24,
    },
  ],
]
`;

exports[`Gravnic calulateNextGameState() Should return the next game state when moving entities up 1`] = `
Array [
  Array [
    Object {
      "entityId": 0,
      "id": 0,
    },
    Object {
      "entityId": 1,
      "id": 1,
    },
    Object {
      "entityId": 1,
      "id": 2,
    },
    Object {
      "entityId": 1,
      "id": 3,
    },
    Object {
      "entityId": 0,
      "id": 4,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 5,
    },
    Object {
      "entityId": 1,
      "id": 6,
    },
    Object {
      "entityId": 2,
      "id": 7,
    },
    Object {
      "entityId": 1,
      "id": 8,
    },
    Object {
      "entityId": 1,
      "id": 9,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 10,
    },
    Object {
      "entityId": 1,
      "id": 11,
    },
    Object {
      "entityId": 1,
      "id": 12,
    },
    Object {
      "entityId": 1,
      "id": 13,
    },
    Object {
      "entityId": 1,
      "id": 14,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 15,
    },
    Object {
      "entityId": 1,
      "id": 16,
    },
    Object {
      "entityId": 1,
      "id": 17,
    },
    Object {
      "entityId": 1,
      "id": 18,
    },
    Object {
      "entityId": 1,
      "id": 19,
    },
  ],
  Array [
    Object {
      "entityId": 0,
      "id": 20,
    },
    Object {
      "entityId": 1,
      "id": 21,
    },
    Object {
      "entityId": 1,
      "id": 22,
    },
    Object {
      "entityId": 1,
      "id": 23,
    },
    Object {
      "entityId": 0,
      "id": 24,
    },
  ],
]
`;

exports[`Gravnic changeGravityDirection() Should return an array of all game states from the move 1`] = `
Array [
  Array [
    Array [
      Object {
        "entityId": 0,
        "id": 0,
      },
      Object {
        "entityId": 1,
        "id": 1,
      },
      Object {
        "entityId": 1,
        "id": 2,
      },
      Object {
        "entityId": 1,
        "id": 3,
      },
      Object {
        "entityId": 0,
        "id": 4,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 5,
      },
      Object {
        "entityId": 1,
        "id": 6,
      },
      Object {
        "entityId": 1,
        "id": 7,
      },
      Object {
        "entityId": 1,
        "id": 8,
      },
      Object {
        "entityId": 1,
        "id": 9,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 10,
      },
      Object {
        "entityId": 1,
        "id": 11,
      },
      Object {
        "entityId": 1,
        "id": 12,
      },
      Object {
        "entityId": 1,
        "id": 13,
      },
      Object {
        "entityId": 1,
        "id": 14,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 15,
      },
      Object {
        "entityId": 1,
        "id": 16,
      },
      Object {
        "entityId": 1,
        "id": 17,
      },
      Object {
        "entityId": 1,
        "id": 18,
      },
      Object {
        "entityId": 1,
        "id": 19,
      },
    ],
    Array [
      Object {
        "entityId": 0,
        "id": 20,
      },
      Object {
        "entityId": 1,
        "id": 21,
      },
      Object {
        "entityId": 2,
        "id": 22,
      },
      Object {
        "entityId": 1,
        "id": 23,
      },
      Object {
        "entityId": 0,
        "id": 24,
      },
    ],
  ],
  Array [
    Array [
      Object {
        "entityId": 0,
        "id": 0,
      },
      Object {
        "entityId": 1,
        "id": 1,
      },
      Object {
        "entityId": 1,
        "id": 2,
      },
      Object {
        "entityId": 1,
        "id": 3,
      },
      Object {
        "entityId": 0,
        "id": 4,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 5,
      },
      Object {
        "entityId": 1,
        "id": 6,
      },
      Object {
        "entityId": 1,
        "id": 7,
      },
      Object {
        "entityId": 1,
        "id": 8,
      },
      Object {
        "entityId": 1,
        "id": 9,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 10,
      },
      Object {
        "entityId": 1,
        "id": 11,
      },
      Object {
        "entityId": 1,
        "id": 12,
      },
      Object {
        "entityId": 1,
        "id": 13,
      },
      Object {
        "entityId": 1,
        "id": 14,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 15,
      },
      Object {
        "entityId": 1,
        "id": 16,
      },
      Object {
        "entityId": 1,
        "id": 17,
      },
      Object {
        "entityId": 1,
        "id": 18,
      },
      Object {
        "entityId": 1,
        "id": 19,
      },
    ],
    Array [
      Object {
        "entityId": 0,
        "id": 20,
      },
      Object {
        "entityId": 1,
        "id": 21,
      },
      Object {
        "entityId": 2,
        "id": 22,
      },
      Object {
        "entityId": 1,
        "id": 23,
      },
      Object {
        "entityId": 0,
        "id": 24,
      },
    ],
  ],
]
`;

exports[`Gravnic changeGravityDirection() Should return an array of all game states from the move 2`] = `
Array [
  Array [
    Array [
      Object {
        "entityId": 0,
        "id": 0,
      },
      Object {
        "entityId": 1,
        "id": 1,
      },
      Object {
        "entityId": 2,
        "id": 2,
      },
      Object {
        "entityId": 1,
        "id": 3,
      },
      Object {
        "entityId": 0,
        "id": 4,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 5,
      },
      Object {
        "entityId": 1,
        "id": 6,
      },
      Object {
        "entityId": 1,
        "id": 7,
      },
      Object {
        "entityId": 1,
        "id": 8,
      },
      Object {
        "entityId": 1,
        "id": 9,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 10,
      },
      Object {
        "entityId": 1,
        "id": 11,
      },
      Object {
        "entityId": 1,
        "id": 12,
      },
      Object {
        "entityId": 1,
        "id": 13,
      },
      Object {
        "entityId": 1,
        "id": 14,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 15,
      },
      Object {
        "entityId": 1,
        "id": 16,
      },
      Object {
        "entityId": 1,
        "id": 17,
      },
      Object {
        "entityId": 1,
        "id": 18,
      },
      Object {
        "entityId": 1,
        "id": 19,
      },
    ],
    Array [
      Object {
        "entityId": 0,
        "id": 20,
      },
      Object {
        "entityId": 1,
        "id": 21,
      },
      Object {
        "entityId": 1,
        "id": 22,
      },
      Object {
        "entityId": 1,
        "id": 23,
      },
      Object {
        "entityId": 0,
        "id": 24,
      },
    ],
  ],
  Array [
    Array [
      Object {
        "entityId": 0,
        "id": 0,
      },
      Object {
        "entityId": 1,
        "id": 1,
      },
      Object {
        "entityId": 2,
        "id": 2,
      },
      Object {
        "entityId": 1,
        "id": 3,
      },
      Object {
        "entityId": 0,
        "id": 4,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 5,
      },
      Object {
        "entityId": 1,
        "id": 6,
      },
      Object {
        "entityId": 1,
        "id": 7,
      },
      Object {
        "entityId": 1,
        "id": 8,
      },
      Object {
        "entityId": 1,
        "id": 9,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 10,
      },
      Object {
        "entityId": 1,
        "id": 11,
      },
      Object {
        "entityId": 1,
        "id": 12,
      },
      Object {
        "entityId": 1,
        "id": 13,
      },
      Object {
        "entityId": 1,
        "id": 14,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 15,
      },
      Object {
        "entityId": 1,
        "id": 16,
      },
      Object {
        "entityId": 1,
        "id": 17,
      },
      Object {
        "entityId": 1,
        "id": 18,
      },
      Object {
        "entityId": 1,
        "id": 19,
      },
    ],
    Array [
      Object {
        "entityId": 0,
        "id": 20,
      },
      Object {
        "entityId": 1,
        "id": 21,
      },
      Object {
        "entityId": 1,
        "id": 22,
      },
      Object {
        "entityId": 1,
        "id": 23,
      },
      Object {
        "entityId": 0,
        "id": 24,
      },
    ],
  ],
  Array [
    Array [
      Object {
        "entityId": 0,
        "id": 0,
      },
      Object {
        "entityId": 1,
        "id": 1,
      },
      Object {
        "entityId": 2,
        "id": 2,
      },
      Object {
        "entityId": 1,
        "id": 3,
      },
      Object {
        "entityId": 0,
        "id": 4,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 5,
      },
      Object {
        "entityId": 1,
        "id": 6,
      },
      Object {
        "entityId": 1,
        "id": 7,
      },
      Object {
        "entityId": 1,
        "id": 8,
      },
      Object {
        "entityId": 1,
        "id": 9,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 10,
      },
      Object {
        "entityId": 1,
        "id": 11,
      },
      Object {
        "entityId": 1,
        "id": 12,
      },
      Object {
        "entityId": 1,
        "id": 13,
      },
      Object {
        "entityId": 1,
        "id": 14,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 15,
      },
      Object {
        "entityId": 1,
        "id": 16,
      },
      Object {
        "entityId": 1,
        "id": 17,
      },
      Object {
        "entityId": 1,
        "id": 18,
      },
      Object {
        "entityId": 1,
        "id": 19,
      },
    ],
    Array [
      Object {
        "entityId": 0,
        "id": 20,
      },
      Object {
        "entityId": 1,
        "id": 21,
      },
      Object {
        "entityId": 1,
        "id": 22,
      },
      Object {
        "entityId": 1,
        "id": 23,
      },
      Object {
        "entityId": 0,
        "id": 24,
      },
    ],
  ],
  Array [
    Array [
      Object {
        "entityId": 0,
        "id": 0,
      },
      Object {
        "entityId": 1,
        "id": 1,
      },
      Object {
        "entityId": 2,
        "id": 2,
      },
      Object {
        "entityId": 1,
        "id": 3,
      },
      Object {
        "entityId": 0,
        "id": 4,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 5,
      },
      Object {
        "entityId": 1,
        "id": 6,
      },
      Object {
        "entityId": 1,
        "id": 7,
      },
      Object {
        "entityId": 1,
        "id": 8,
      },
      Object {
        "entityId": 1,
        "id": 9,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 10,
      },
      Object {
        "entityId": 1,
        "id": 11,
      },
      Object {
        "entityId": 1,
        "id": 12,
      },
      Object {
        "entityId": 1,
        "id": 13,
      },
      Object {
        "entityId": 1,
        "id": 14,
      },
    ],
    Array [
      Object {
        "entityId": 1,
        "id": 15,
      },
      Object {
        "entityId": 1,
        "id": 16,
      },
      Object {
        "entityId": 1,
        "id": 17,
      },
      Object {
        "entityId": 1,
        "id": 18,
      },
      Object {
        "entityId": 1,
        "id": 19,
      },
    ],
    Array [
      Object {
        "entityId": 0,
        "id": 20,
      },
      Object {
        "entityId": 1,
        "id": 21,
      },
      Object {
        "entityId": 1,
        "id": 22,
      },
      Object {
        "entityId": 1,
        "id": 23,
      },
      Object {
        "entityId": 0,
        "id": 24,
      },
    ],
  ],
]
`;

exports[`Gravnic convertTilesToGameState() Converts the tiles into a starting game state 1`] = `
Array [
  Array [
    Object {
      "entityId": 0,
      "id": 0,
    },
    Object {
      "entityId": 1,
      "id": 1,
    },
    Object {
      "entityId": 1,
      "id": 2,
    },
    Object {
      "entityId": 1,
      "id": 3,
    },
    Object {
      "entityId": 0,
      "id": 4,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 5,
    },
    Object {
      "entityId": 1,
      "id": 6,
    },
    Object {
      "entityId": 1,
      "id": 7,
    },
    Object {
      "entityId": 1,
      "id": 8,
    },
    Object {
      "entityId": 1,
      "id": 9,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 10,
    },
    Object {
      "entityId": 1,
      "id": 11,
    },
    Object {
      "entityId": 2,
      "id": 12,
    },
    Object {
      "entityId": 1,
      "id": 13,
    },
    Object {
      "entityId": 1,
      "id": 14,
    },
  ],
  Array [
    Object {
      "entityId": 1,
      "id": 15,
    },
    Object {
      "entityId": 1,
      "id": 16,
    },
    Object {
      "entityId": 1,
      "id": 17,
    },
    Object {
      "entityId": 1,
      "id": 18,
    },
    Object {
      "entityId": 1,
      "id": 19,
    },
  ],
  Array [
    Object {
      "entityId": 0,
      "id": 20,
    },
    Object {
      "entityId": 1,
      "id": 21,
    },
    Object {
      "entityId": 1,
      "id": 22,
    },
    Object {
      "entityId": 1,
      "id": 23,
    },
    Object {
      "entityId": 0,
      "id": 24,
    },
  ],
]
`;
